src/time_integration.jl

```diff
@@ -1,4 +1,665 @@
+        copyto!(u,u0)
+        copyto!(v,v0)
+        copyto!(η,η0)
+    end
+
+    # finalise feedback and output
+    feedback_end!(feedback)
+    output_close!(netCDFfiles,feedback,S)
+
+    if S.parameters.return_time
+        return feedback.tend - feedback.t0
+    else
+        return PrognosticVars{Tprog}(remove_halo(u,v,η,sst,S)...)
+    end
+end
+
+"""Integrate ShallowWaters forward and checkpoints the integration. This is the
+integration function needed if we want to use Enzyme on the model."""
+function checkpointed_integration(S, scheme)
+
+    # setup
+    Diag = S.Diag
+    Prog = S.Prog
+
+    @unpack u,v,η,sst = Prog
+    @unpack u0,v0,η0 = Diag.RungeKutta
+    @unpack u1,v1,η1 = Diag.RungeKutta
+    @unpack du,dv,dη = Diag.Tendencies
+    @unpack du_sum,dv_sum,dη_sum = Diag.Tendencies
+    @unpack du_comp,dv_comp,dη_comp = Diag.Tendencies
+
+    @unpack um,vm = Diag.SemiLagrange
+
+    @unpack dynamics,RKo,RKs,tracer_advection = S.parameters
+    @unpack time_scheme,compensated = S.parameters
+    @unpack RKaΔt,RKbΔt = S.constants
+    @unpack Δt_Δ,Δt_Δs = S.constants
+
+    @unpack nt,dtint = S.grid
+    @unpack nstep_advcor,nstep_diff,nadvstep,nadvstep_half = S.grid
+
+    # calculate layer thicknesses for initial conditions
+    ShallowWaters.thickness!(Diag.VolumeFluxes.h,η,S.forcing.H)
+    ShallowWaters.Ix!(Diag.VolumeFluxes.h_u,Diag.VolumeFluxes.h)
+    ShallowWaters.Iy!(Diag.VolumeFluxes.h_v,Diag.VolumeFluxes.h)
+    ShallowWaters.Ixy!(Diag.Vorticity.h_q,Diag.VolumeFluxes.h)
+
+    # calculate PV terms for initial conditions
+    urhs = convert(Diag.PrognosticVarsRHS.u,u)
+    vrhs = convert(Diag.PrognosticVarsRHS.v,v)
+    ηrhs = convert(Diag.PrognosticVarsRHS.η,η)
+
+    ShallowWaters.advection_coriolis!(urhs,vrhs,ηrhs,Diag,S)
+    ShallowWaters.PVadvection!(Diag,S)
+
+    # propagate initial conditions
+    copyto!(u0,u)
+    copyto!(v0,v)
+    copyto!(η0,η)
+
+    # store initial conditions of sst for relaxation
+    copyto!(Diag.SemiLagrange.sst_ref,sst)
+
+    # adding causes bug in Enzyme, don't really need these functions
+    # (or at least it doesn't seem like I do)
+    # feedback, output initialisation and storing initial conditions
+    # feedback = feedback_init(S)
+    # netCDFfiles = ShallowWaters.NcFiles(feedback,S)
+    # ShallowWaters.output_nc!(0,netCDFfiles,Prog,Diag,S)
+
+    # nans_detected = false
+    t = 0                       # model time
+    # run integration loop with checkpointing
+    loop(S, scheme)
+
+    # end
+
+    # separately defined J
+    return S.parameters.J
+
+    # final state cost function
+    # return S.Prog.η[24,24]
+
+    # spatially averaged energy cost function
+    # temp = ShallowWaters.PrognosticVars{Float32}(ShallowWaters.remove_halo(u,v,η,sst,S)...)
+    # S.parameters.J = (sum(temp.u.^2) + sum(temp.v.^2)) / (S.grid.nx * S.grid.ny)
+    # return S.parameters.J
+
+end
+
+function loop(S,scheme)
+
+    @checkpoint_struct scheme S for S.parameters.i = 1:S.grid.nt
```

Comment:
This is the problem line. If I don't add Checkpointing then running `include("./src/ShallowWaters.jl")` will fail because `@checkpoint_struct` will be undefined. In response to a more recent comment, I might still be able to integrate the function as normal with this, but I haven't tried this myself and it might also still run checkpointing? I'm not sure what exactly will change

Reply:
I'm happy to find an alternative place to put this function though! Maybe create it's own script and instruct users to include that script separately if they want to checkpoint?

Reply:
Then in `time_integration.jl` I'd leave the single definition for model integration